### PR TITLE
Last lsp changes needed

### PR DESCRIPTION
**Pull request by prozacchiwawa**
Several odd changes:
1) Pulling out the standard macros and primitives so they can be used in other CompilerOpts providers.
2) Add a liveness check switch that allows the frontend to not check for liveness when computing the helpers (needed for highlighting).
3) The frontend output now remembers what was included and provides consumers with the ability to inspect these forms at the source level.
4) fix: previously forms with fewer than 3 elements were assumed to need to be wrapped in mod, but degenerate mod forms caused this to recurse if they were degenerate mod forms.  Now only follow this path without trying to find a mod token if the toplevel form is a singular empty unit token.

### CODE REVIEW

File: src/compiler/clvm.rs [hunk 1]
```diff
@@ -631,9 +631,9 @@ pub fn sha256tree(s: Rc<SExp>) -> Vec<u8> {
...
             .clone()
         }
-        SExp::Nil(_) => sha256tree_from_atom(vec![]),
-        SExp::Integer(_, i) => sha256tree_from_atom(u8_from_number(i.clone())),
-        SExp::QuotedString(_, _, v) => sha256tree_from_atom(v.clone()),
-        SExp::Atom(_, v) => sha256tree_from_atom(v.clone()),
+        SExp::Nil(_) => sha256tree_from_atom(&[]),
+        SExp::Integer(_, i) => sha256tree_from_atom(&u8_from_number(i.clone())),
+        SExp::QuotedString(_, _, v) => sha256tree_from_atom(&v.clone()),
+        SExp::Atom(_, v) => sha256tree_from_atom(&v.clone()),
```
Comment (arvidn) [inline src/compiler/clvm.rs]: it looks like these calls to `clone()` are redundant, since you take a slice now.

File: src/compiler/clvm.rs [hunk 2]
```diff
@@ -608,10 +608,10 @@ pub fn parse_and_run(
     }
 }
 
-fn sha256tree_from_atom(v: Vec<u8>) -> Vec<u8> {
+pub fn sha256tree_from_atom(v: &[u8]) -> Vec<u8> {
     sha256(
         Bytes::new(Some(BytesFromType::Raw(vec![1])))
-            .concat(&Bytes::new(Some(BytesFromType::Raw(v)))),
+            .concat(&Bytes::new(Some(BytesFromType::Raw(v.to_vec())))),
```
Comment (arvidn) [inline src/compiler/clvm.rs]: it looks quite strange to heap-allocate and copy the bytes here, just to concatenate and run sha-256 over them. That should be possible without any heap allocation or copying (and possibly less code).

File: src/compiler/clvm.rs (hunk 1 above)
Comment (prozacchiwawa) [reply src/compiler/clvm.rs]: yes you're right thanks.

File: src/compiler/clvm.rs (hunk 2 above)
Comment (prozacchiwawa) [reply src/compiler/clvm.rs]: slicked these up and added a test.

File: src/tests/compiler/clvm.rs [hunk 3]
```diff
@@ -213,3 +214,19 @@ fn test_random_int_just_the_conversion_functions_and_no_other_things_from_the_st
...
+#[test]
+fn test_sha256_tree_hash() {
+    let filename = "resources/tests/assert.clvm.out";
+    let assert_clvm_compiled = fs::read_to_string(filename).expect("should exist");
+    let parsed = parse_sexp(
+        Srcloc::start(filename),
+        assert_clvm_compiled.as_bytes().iter().copied(),
+    )
+    .expect("should parse");
+    let hash_result = Bytes::new(Some(BytesFromType::Raw(sha256tree(parsed[0].clone())))).hex();
```
Comment (arvidn) [inline src/tests/compiler/clvm.rs:227]: it seems to me this could be simplified

Comment (arvidn) [inline src/tests/compiler/clvm.rs:227]:
```suggestion
    let hash_result = sha256tree(&parsed[0]);
```

File: src/tests/compiler/clvm.rs [hunk 4]
```diff
@@ -213,3 +214,19 @@ fn test_random_int_just_the_conversion_functions_and_no_other_things_from_the_st
...
+    assert_eq!(
+        hash_result,
+        "156e86309040ed6bbfee805c9c6ca7eebc140490bd1b97d6d18fb8ebc91fd05a"
```
Comment (arvidn) [inline src/tests/compiler/clvm.rs:230]: it's not so easy to tell that this is correct by inspecting the test

**Review by arvidn — APPROVED** [review]

File: src/tests/compiler/clvm.rs (hunk 4 above)
Comment (prozacchiwawa) [reply src/tests/compiler/clvm.rs:230]: Hmm true.  I do wonder how to communicate things like this.

File: src/tests/compiler/clvm.rs (hunk 3 above)
Comment (prozacchiwawa) [reply src/tests/compiler/clvm.rs:227]: It's an Rc<SExp> so we can't pass down a reference here (and really shouldn't).

Comment (prozacchiwawa) [reply src/tests/compiler/clvm.rs:227]: I'll give some thought to simplifying the test cases.

**Pull request merged**